lib/puppet/parser/resource.rb

```diff
@@ -94,8 +94,18 @@ def exported=(value)
     end
   end
 
-  # Do any finishing work on this object, called before evaluation or
-  # before storage/translation. The method does nothing the second time
+  # Finish the evaluation by assigning defaults and scope tags
+  # @api private
+  #
+  def finish_evaluation
+    return if @evaluation_finished
+    add_defaults
+    add_scope_tags
+    @evaluation_finished = true
```

Comment:
Although I think the most likely step to raise an exception is `validate`, which is not part of the `finish_evaluation` scope